federation/schema_syncer_test.go

```diff
@@ -243,3 +243,113 @@ func TestExecutorQueriesWithCustomSchemaSyncer(t *testing.T) {
 	// Run the same query and validate that it works
 	runAndValidateQueryResults(t, ctx, e, query2, expectedOutput2)
 }
+
+func TestOnlyShadowServiceKnowsAboutNewField(t *testing.T) {
+	type User struct {
+		Id          int64
+		OrgId       int64
+		Name        string
+		Email       string
+		PhoneNumber string
+	}
+	s1 := schemabuilder.NewSchemaWithName("schema1")
+	user := s1.Object("User", User{}, schemabuilder.RootObject)
+	user.Key("id")
+	s1.Query().FieldFunc("users", func(ctx context.Context) ([]*User, error) {
+		users := make([]*User, 0, 1)
+		users = append(users, &User{Id: int64(1), OrgId: int64(1), Name: "testUser", Email: "email@gmail.com", PhoneNumber: "555-5555"})
+		users = append(users, &User{Id: int64(2), OrgId: int64(2), Name: "testUser2", Email: "email@gmail.com", PhoneNumber: "555-5555"})
+		return users, nil
+	})
+
+	type UserWithContactInfo struct {
+		Id    int64
+		OrgId int64
+		Name  string
+	}
+
+	s2 := schemabuilder.NewSchemaWithName("schema2")
+	s2.FederatedFieldFunc("User", func(args struct{ Keys []*UserWithContactInfo }) []*UserWithContactInfo {
+		return args.Keys
+	})
+	userWithContactInfo := s2.Object("User", UserWithContactInfo{})
+	userWithContactInfo.FieldFunc("isCool", func(ctx context.Context) (bool, error) { return true, nil })
+
+	ctx := context.Background()
+	execs, err := makeExecutors(map[string]*schemabuilder.Schema{
+		"schema1": s1,
+		"schema2": s2,
+	})
+	require.NoError(t, err)
+
+	// Write the schemas to a file
+	services := []string{"schema1", "schema2"}
+	for _, service := range services {
+		schema, err := fetchSchema(ctx, execs[service], nil)
+		require.NoError(t, err)
+		err = writeSchemaToFile(service, schema.Result)
+		require.NoError(t, err)
+	}
+
+	// Creata file schema syncer that reads the schemas from the
+	// written files and listens to updates if those change
+	schemaSyncer := newFileSchemaSyncer(ctx, services)
+	e, err := NewExecutor(ctx, execs, &CustomExecutorArgs{
+		SchemaSyncer:              schemaSyncer,
+		SchemaSyncIntervalSeconds: func(ctx context.Context) int64 { return 100 },
+	})
+	require.NoError(t, err)
+
+	query := `query Foo {
+					users {
+						isCool
+					}
+				}`
+	expectedOutput := `{
+					"users":[
+						{
+							"__key":1, 
+							"isCool":true
+						},
+						{
+							"__key":2,
+							"isCool":true
+						}
+					]
+				}`
+
+	// Run a federated query and ensure that it works
+	runAndValidateQueryResults(t, ctx, e, query, expectedOutput)
+
+	type UserWithContactInfoNew struct {
+		Id    int64
+		OrgId int64
+		Name  string
+		// When the gateway knows about email, it will fill it correctly
+		Email string `graphql:",optional"`
```

Comment:
huh, could you point me the code where we deal with the optional field?

Reply:
Yup! https://github.com/samsarahq/thunder/blob/master/graphql/schemabuilder/input.go#L160